crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1908 @@
+    while (--rank > 0)
+        scalar_mult_add(out, ++lhs, ++rhs);
+}
+
+/* In-place NTT transform of a vector */
+static void vector_ntt(scalar *a, int rank)
+{
+    while (rank-- > 0)
+        scalar_ntt(a++);
+}
+
+/* In-place inverse NTT transform of a vector */
+static void vector_inverse_ntt(scalar *a, int rank)
+{
+    while (rank-- > 0)
+        scalar_inverse_ntt(a++);
+}
+
+/* Here, the output vector must not overlap with the inputs */
+static void
+matrix_mult(scalar *out, const scalar *m, const scalar *a, int rank)
+{
+    const scalar *ar;
+    int i, j;
+
+    for (i = rank; i-- > 0; ++out) {
+        scalar_mult(out, m++, ar = a);
+        for (j = rank - 1; j > 0; --j)
+            scalar_mult_add(out, m++, ++ar);
+    }
+}
+
+/* Here, the output vector must not overlap with the inputs */
+static void
+matrix_mult_transpose(scalar *out, const scalar *m, const scalar *a, int rank)
+{
+    const scalar *mc = m, *mr, *ar;
+    int i, j;
+
+    for (i = rank; i-- > 0; ++out)  {
+        scalar_mult(out, mr = mc++, ar = a);
+        for (j = rank; --j > 0; )
+            scalar_mult_add(out, (mr += rank), ++ar);
+    }
+}
+
+/*-
+ * Expands the matrix from a seed for key generation and for encaps-CPA.
+ * NOTE: FIPS 203 matrix "A" is the transpose of this matrix, computed
+ * by appending the (i,j) indices to the seed in the opposite order!
+ *
+ * Where FIPS 203 computes t = A * s + e, we use the transpose of "m".
+ */
+static __owur
+int matrix_expand(EVP_MD_CTX *mdctx, ML_KEM_KEY *key)
+{
+    scalar *out = key->m;
+    uint8_t input[ML_KEM_RANDOM_BYTES + 2];
+    int rank = key->vinfo->rank;
+    int i, j;
+
+    memcpy(input, key->rho, ML_KEM_RANDOM_BYTES);
+    for (i = 0; i < rank; i++) {
+        for (j = 0; j < rank; j++) {
+            input[ML_KEM_RANDOM_BYTES] = i;
+            input[ML_KEM_RANDOM_BYTES + 1] = j;
+            if (!EVP_DigestInit_ex(mdctx, key->shake128_md, NULL)
+                || !EVP_DigestUpdate(mdctx, input, sizeof(input))
+                || !sample_scalar(out++, mdctx))
+                return 0;
+        }
+    }
+    return 1;
+}
+
+/*
+ * Algorithm 7 from the spec, with eta fixed to two and the PRF call
+ * included. Creates binominally distributed elements by sampling 2*|eta| bits,
+ * and setting the coefficient to the count of the first bits minus the count of
+ * the second bits, resulting in a centered binomial distribution. Since eta is
+ * two this gives -2/2 with a probability of 1/16, -1/1 with probability 1/4,
+ * and 0 with probability 3/8.
+ */
+static __owur
+int cbd_2(scalar *out, uint8_t in[ML_KEM_RANDOM_BYTES + 1],
+          EVP_MD_CTX *mdctx, const ML_KEM_KEY *key)
+{
+    uint16_t *curr = out->c, *end = curr + DEGREE;
+    uint8_t randbuf[4 * DEGREE / 8], *r = randbuf;  /* 64 * eta slots */
+    uint16_t value, mask;
+    uint8_t b;
+
+    if (!prf(randbuf, sizeof(randbuf), in, mdctx, key))
+        return 0;
+
+#define bit0(b) (b & 1)
+#define bitn(n, b) ((b >> n) & 1)
```

Comment:
I think it would make sense for these to be up the top as general utility macros.
Defining them in the middle of this function and using them in the next isn't desirable IMO.

Reply:
Done, and put them to use in a couple more places.